doc/cimple.md

```diff
@@ -0,0 +1,452 @@
+Only matches sequences of `#define`s longer than 5 to avoid some false positives.
+Also, the sequence must have a common prefix of at least 2 components. I.e.
+`AAA` is not a sufficient common prefix, but `AAA_BBB` is.
+Lastly, we only require enums for small-int enums, i.e. all enumerators have a
+constant int expression value less than or equal to 255.
+
+**Reason:** `enum` constants are safer, and can potentially be type-checked
+more thoroughly.
+
+## `-Wenum-from-int` (global)
+
+Checks that `_from_int` functions for `enum`s are complete.
+
+**Reason:** ensures that no enumerators are missed in conversion functions that
+turn `int`s into `enum`s. Type-cast is not permitted, because some values of
+type `int` are not in the enumeration.
+
+## `-Wenum-names`
+
+Checks that `enum` value constants have the same prefix as the `enum` type,
+except they should be SCREAMING_CASE instead of Camel_Snake. There are currently
+37 exemptions to this rule. New enums should follow the naming convention.
+
+**Reason:** this naming convention helps identify the type of an `enum` constant
+at first glance.
+
+## `-Wenum-to-string` (global)
+
+Checks that `_to_string` functions for `enum`s are complete.
+
+**Reason:** we provide `to_string` functions for `enum` but don't want to
+manually maintain them. This linter checks that the function is exactly what
+we want it to be, and the error message will say what the function should look
+like.
+
+## `-Wenum-unpack` (global)
+
+Checks that `_unpack` functions for `enum`s are complete.
+
+**Reason:** we provide `unpack` functions for `enum` but don't want to
+manually maintain them. This linter checks that the function is exactly what
+we want it to be, and the error message will say what the function should look
+like.
+
+## `-Wfunc-prototypes`
+
+Checks that empty parameter lists in C functions are written as `(void)`.
+
+**Reason:** old-style empty parameter lists written as `()` are risky, because
+C interprets them as variadic. GCC warns about this but sometimes misses one.
+
+## `-Wfunc-scopes`
+
+Checks that static function definitions are marked with `static`.
+
+In C, a function is `static` even if the definition doesn't use `static`, but
+there happens to be another declaration of the function which does.
+
+**Reason:** static/extern qualification of functions should be visible locally.
+It takes mental effort otherwise to look up the declaration to check for storage
+qualifiers.
+
+## `-Wglobal-funcs`
+
+Checks that no extern functions are declared in .c files.
+
+Extern functions must only be declared in .h files. In .c files all declarations
+must be static.
+
+**Reason:** extern declarations in .c files mean that we depend on a function
+not declared in a .h file we can include. This means we're depending on an
+unexported implementation detail, and there is no compiler that can check
+whether our declaration matches the implementation's definition.
+
+## `-Wlarge-struct-params`
```

Comment:
clang-tidy has https://clang.llvm.org/extra/clang-tidy/checks/performance/unnecessary-value-param.html which should be able to detect this.

Reply:
Not quite:

> The check is only applied to parameters of types that are expensive to copy which means they are not trivially copyable or have a non-trivial copy constructor or destructor.

All our types are trivially copyable. That doesn't mean they are cheap to copy.